src/OrchardCore/OrchardCore.Users.Abstractions/Handlers/UpdateUserContext.cs

```diff
@@ -0,0 +1,77 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json.Nodes;
+using OrchardCore.Users.Models;
+
+namespace OrchardCore.Users.Handlers
+{
+    /// <summary>
+    /// Represents a <see cref="UserContextBase"/> for updated user roles.
+    /// </summary>
+    public class UpdateUserContext : UserContextBase
+    {
+        /// <summary>
+        /// Creates a new instance of <see cref="UpdateUserContext"/>.
+        /// </summary>
+        /// <param name="user">The <see cref="IUser"/>.</param>
+        /// <param name="loginProvider">The login provider.</param>
+        /// <param name="externalClaims">The user claims.</param>
+        /// <param name="userProperties">The user properties.</param>
+        public UpdateUserContext(IUser user, string loginProvider, IEnumerable<SerializableClaim> externalClaims, JsonObject userProperties) : base(user)
+        {
+            ExternalClaims = externalClaims.AsEnumerable();
+            LoginProvider = loginProvider;
+            UserProperties = userProperties.DeepClone() as JsonObject;
+        }
+
+        /// <summary>
+        /// Gets the login provider.
+        /// </summary>
+        public string LoginProvider { get; }
+
+        /// <summary>
+        /// Gets a list of external claims.
+        /// </summary>
+        public IEnumerable<SerializableClaim> ExternalClaims { get; }
+
+        /// <summary>
+        /// Gets the user's roles.
+        /// </summary>
+        public IEnumerable<string> UserRoles { get; set; }
+
+        /// <summary>
+        /// Gets the user's roles.
+        /// </summary>
+        public IEnumerable<UserClaim> UserClaims { get; set; }
+
+        /// <summary>
+        /// Gets the user properties.
+        /// </summary>
+        public JsonObject UserProperties { get; }
```

Comment:
Hmm, can we make this class more consistent concerning the property setters? There are several properties get/set, some are get/init etc... Looks a little bit messy to me 😉

Reply:
updated